matrix-e2e.c

```diff
@@ -281,3 +283,141 @@ static int matrix_restore_e2e_account(MatrixConnectionData *conn)
     return ret;
 }
 
+static void key_upload_callback(MatrixConnectionData *conn,
+                                gpointer user_data,
+                                struct _JsonNode *json_root,
+                                const char *body,
+                                size_t body_len, const char *content_type)
+{
+  // TODO
+}
+
+/*
+ * Get a set of device keys for ourselves.  Either by retreiving it from our store
+ * or by generating a new set.
+ *
+ * Returns: 0 on success
+ */
+int matrix_e2e_get_device_keys(MatrixConnectionData *conn, const gchar *device_id)
+{
+    PurpleConnection *pc = conn->pc;
+    JsonObject * json_dev_keys = NULL;
+    OlmAccount *account = olm_account(g_malloc0(olm_account_size()));
+    char *pickled_account = NULL;
+    int ret = 0;
+
+    if (!conn->e2e) {
+        conn->e2e = g_new0(MatrixE2EData,1);
```

Comment:
Done; I've added a matrix_e2e_cleanup_connection and called it from matrix_connection_free.